[Phase 3] HEAD-SHA freshness safeguard so no tier reviews stale pre-fed context

## Story
As a pipeline maintainer,
I want verify the PR head SHA still matches the pre-fed context's stamped SHA before the expensive agentic tiers consume it, and invalidate/re-fetch when a force-push moved HEAD,
so that the pre-agentic optimization can never cause a review of stale code, neutralizing the staleness objection raised in the idea's adversarial review.

## Acceptance Criteria
1. Before the deep/audit/single tiers consume pre-fed context, the cascade cheaply re-checks the PR's current head SHA against the SHA stamped on the pre-fed files (Story 2).
2. If the SHA still matches, the pre-fed context is used as-is (no extra fetch).
3. If HEAD moved (force-push / new push mid-run), the stale pre-fed context is discarded and the run either re-fetches fresh context at the new SHA or safely skips (exit-100 sentinel) to be retried — never proceeds on stale data.
4. The freshness check is bounded to a single lightweight `gh pr view --json headRefOid` call (not a full metadata/diff re-fetch) and is a no-op when the prefetch flag is off.
5. A bats test covers matched-SHA (use cached), moved-SHA (invalidate), and flag-off (no-op) cases.

## Tasks / Subtasks
- [ ] Add a cheap head-SHA re-check before agentic-tier consumption (AC: #1, #4)
  - [ ] Single `gh pr view --json headRefOid` compared to the stamped SHA
- [ ] Invalidate + re-fetch or safe-skip on SHA mismatch (AC: #3)
  - [ ] Re-fetch fresh context at new HEAD, or emit the exit-100 skip sentinel for retry
- [ ] Keep it a no-op when the flag is off (AC: #4)
- [ ] Add bats coverage for matched / moved / flag-off (AC: #5)

## Dev Notes
- The idea's adversarial rebuttal proposes exactly this: the cheap tier validates that HEAD still matches what was pre-fetched before the expensive deep review runs. Implement it as a shell check in `review-one-pr.sh` between the prefetch and the tier-2 launch (~line 874), not inside a prompt.
- `PR_HEAD_SHA` is already resolved once at ~line 108 via `gh pr view --json headRefOid,...`; reuse `compute_ci_status`-style single-call patterns. The freshness call must be minimal (`--json headRefOid` only).
- On mismatch, prefer re-fetch at the new SHA (cheap, keeps the review useful) but the exit-100 skip sentinel is an acceptable safe fallback — the scheduler re-runs and the idempotency marker prevents duplicate reviews (`review-one-pr.sh` idempotency block ~line 318).
- Within a single review-one-pr.sh invocation the tiers run seconds apart in one process, so real staleness is rare — but this closes the window deterministically and is the safeguard that lets Story 6 recommend default-on.
- Depends only on Story 2's SHA-stamped files; independent of the prompt rewires (Stories 3-4). Testing via `tests/` bats wired into lint.yml.

### Project Structure Notes
Localized to `scripts/review-one-pr.sh` (optionally the Story 2 helper lib). New bats test under `tests/`, listed in `.github/workflows/lint.yml`.

### References
- scripts/review-one-pr.sh#head-sha
- scripts/review-one-pr.sh#idempotency
- scripts/lib/ci-status.sh

### Likely target surface
- `scripts/review-one-pr.sh`
- `.github/workflows/lint.yml`

_Story prepared by the BMAD Scrum Master (Bob) for epic #1101. Status: ready-for-dev._

## Dev-Lead: Implementation Complete

### Plan Execution
- [x] `scripts/lib/pr-context-prefetch.sh` — added `assert_prefetch_context_fresh <pr_url> [out_dir]`: flag-gated no-op; reads the SHA stamped on the pre-fed files (metadata `.pr_head_sha`, falling back to the diff `# PR_HEAD_SHA:` header); one bounded `gh pr view --json headRefOid`; returns 0 on match, invalidates files + unsets `PR_CONTEXT_*` exports and returns 100 on mismatch; degrades to 0 on gh failure.
- [x] `scripts/review-one-pr.sh` — call the helper right after triage (`~line 795`) and before the single/deep/audit branch; on return 100 emit the skip JSON (`reason":"head-moved-stale-context`) and `exit 100`. Wrapped in the `PREFETCH_CONTEXT_ENABLED` gate for a byte-identical no-op when off.
- [x] `.github/workflows/lint.yml` — added `tests/test_prefetch_freshness.bats` to the `bats` job list.
- [x] `tests/test_prefetch_freshness.bats` — matched / moved / flag-off (AC #5) plus nothing-prefed and gh-failure degrade cases.

### Test Results
```
1..6
ok 1 freshness OFF (unset): no-op, no gh call
ok 2 freshness OFF (explicit false): no-op, no gh call
ok 3 freshness ON matched: returns 0 and keeps the pre-fed context
ok 4 freshness ON moved: invalidates stale context and returns 100
ok 5 freshness ON but nothing pre-fed: no-op returns 0, no gh call
ok 6 freshness ON gh failure: degrades to 0 and keeps the pre-fed context
```

Also green: repo lint (`.dev-lead/scripts/dev-lead-lint.sh` — shellcheck on 80 files + agent-profile validation), `shellcheck` on both changed scripts (only pre-existing SC1091 source-follow info), and the full `lint.yml` bats list with no regressions.

### Files Changed
- `scripts/lib/pr-context-prefetch.sh`: new `assert_prefetch_context_fresh` freshness-check helper.
- `scripts/review-one-pr.sh`: invoke the freshness check between triage and the agentic tiers; safe-skip (exit 100) on a moved HEAD.
- `.github/workflows/lint.yml`: register the new bats file.
- `tests/test_prefetch_freshness.bats`: new coverage.

### Notes
- Placed the single check after triage and before the branch so it covers all three agentic consumers (single / deep / audit) with one `gh pr view --json headRefOid` — triage itself does not consume the pre-fed context.
- On a moved HEAD the run takes the **safe-skip (exit 100)** fallback (AC #3's accepted alternative to in-process re-fetch): the scheduler re-runs, resolves the new head SHA, re-prefetches, and the idempotency marker (no marker is written on this skip) prevents a duplicate review.
- gh failure on the check degrades to proceed on the already-SHA-bound context rather than crashing, avoiding a skip/retry storm on a transient blip.

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -99,6 +99,7 @@ jobs:
             tests/test_safety_checks_triage_inline.bats \
             tests/test_prefetch_context.bats \
             tests/test_prefetch_prompt_rewire.bats \
+            tests/test_prefetch_freshness.bats \
             tests/test_engine_unavailable_notice.bats \
             tests/test_batch_skip_reporting.bats \
             tests/test_review_one_pr_ci_pending.bats \

--- a/scripts/lib/pr-context-prefetch.sh
+++ b/scripts/lib/pr-context-prefetch.sh
@@ -96,3 +96,68 @@ prefetch_pr_context() {
   [ "$meta_stamped" = true ] && export PR_CONTEXT_METADATA_FILE="$meta_file"
   return 0
 }
+
+# assert_prefetch_context_fresh <pr_url> [out_dir]
+#
+#   HEAD-SHA freshness safeguard (epic #1101, Story 5 / #1106). Run ONCE just
+#   before the agentic tiers (deep / audit / single) consume the pre-fed
+#   context, this cheaply re-checks the PR's CURRENT head SHA against the SHA
+#   prefetch_pr_context stamped on the pre-fed files — closing the window where a
+#   force-push / new push mid-run could otherwise let a tier review stale code.
+#
+#   Fetch discipline (AC #4): bounded to a SINGLE lightweight
+#   `gh pr view --json headRefOid` — never a full metadata/diff re-fetch.
+#
+#   Gating (AC #4): a no-op returning 0 (doing NO gh call) unless
+#   PREFETCH_CONTEXT_ENABLED=true. Also a no-op when no pre-fed context was
+#   exported (no stamp to validate).
+#
+#   Returns:
+#     0   — flag off, nothing pre-fed, OR current head SHA still matches the
+#           stamp (AC #2: pre-fed context is fresh — used as-is, no extra fetch).
+#     100 — HEAD moved (AC #3): the stale pre-fed context is INVALIDATED (files
+#           removed, PR_CONTEXT_* exports unset) so no tier can consume it; the
+#           caller takes the exit-100 skip sentinel to retry at the new SHA (the
+#           idempotency marker prevents a duplicate review).
+#
+#   A gh failure on the check degrades to 0 (proceed on the existing stamped
+#   context) rather than crashing: the context is already SHA-bound and this is a
+#   belt-and-suspenders safeguard, so a transient gh blip must not force a
+#   skip/retry storm.
+assert_prefetch_context_fresh() {
+  [ "${PREFETCH_CONTEXT_ENABLED:-false}" = "true" ] || return 0
+
+  local pr_url="${1:?assert_prefetch_context_fresh: pr_url required}"
+  local out_dir="${2:-/tmp/cascade}"
+
+  local meta_file="${PR_CONTEXT_METADATA_FILE:-$out_dir/pr-context-metadata.json}"
+  local diff_file="${PR_CONTEXT_DIFF_FILE:-$out_dir/pr-context-diff.txt}"
+
+  # Resolve the SHA stamped on the pre-fed context: prefer the metadata file's
+  # top-level .pr_head_sha, fall back to the diff header. Neither present => no
+  # pre-fed context to validate, so no-op (no gh call).
+  local stamped_sha=""
+  if [ -f "$meta_file" ]; then
+    stamped_sha=$(jq -r '.pr_head_sha // empty' "$meta_file" 2>/dev/null || true)
+  fi
+  if [ -z "$stamped_sha" ] && [ -f "$diff_file" ]; then
+    stamped_sha=$(sed -n 's/^# PR_HEAD_SHA: //p' "$diff_file" 2>/dev/null | head -1)
+  fi
+  [ -n "$stamped_sha" ] || return 0
+
+  # The single bounded freshness call (AC #4). An inconclusive fetch degrades to
+  # proceed on the existing stamp rather than crashing.
+  local snapshot current_sha
+  snapshot=$(gh pr view "$pr_url" --json headRefOid 2>/dev/null) || return 0
+  current_sha=$(printf '%s' "$snapshot" | jq -r '.headRefOid // empty' 2>/dev/null || true)
+  [ -n "$current_sha" ] || return 0
+
+  # Matched (AC #2): pre-fed context is fresh — use as-is.
+  [ "$current_sha" = "$stamped_sha" ] && return 0
+
+  # Moved (AC #3): discard the stale pre-fed context so no tier can consume it,
+  # and signal the caller to take the exit-100 skip path.
+  rm -f "$diff_file" "$meta_file" 2>/dev/null || true
+  unset PR_CONTEXT_DIFF_FILE PR_CONTEXT_METADATA_FILE
+  return 100
+}

--- a/scripts/lib/pr-context-prefetch.sh
+++ b/scripts/lib/pr-context-prefetch.sh
@@ -141,7 +141,7 @@ assert_prefetch_context_fresh() {
     stamped_sha=$(jq -r '.pr_head_sha // empty' "$meta_file" 2>/dev/null || true)
   fi
   if [ -z "$stamped_sha" ] && [ -f "$diff_file" ]; then
-    stamped_sha=$(sed -n 's/^# PR_HEAD_SHA: //p' "$diff_file" 2>/dev/null | head -1)
+    stamped_sha=$(sed -n '/^# PR_HEAD_SHA: /{s/^# PR_HEAD_SHA: //p;q;}' "$diff_file" 2>/dev/null || true)
   fi
   [ -n "$stamped_sha" ] || return 0
 

--- a/scripts/review-one-pr.sh
+++ b/scripts/review-one-pr.sh
@@ -793,6 +793,26 @@ TRIAGE_RISK=$(echo "$TRIAGE_RESULT" | jq -r '.risk')
 TRIAGE_SIGNALS=$(echo "$TRIAGE_RESULT" | jq -r '.signals | join(", ")')
 echo "    [tier1] escalate=$TRIAGE_ESCALATE risk=$TRIAGE_RISK signals=[$TRIAGE_SIGNALS]"
 
+# HEAD-SHA freshness safeguard (epic #1101, Story 5 / #1106). Just before the
+# agentic tiers (single / deep / audit) consume the Story 2 pre-fed context,
+# cheaply re-check the PR's CURRENT head SHA against the SHA that prefetch
+# stamped on the pre-fed files — a single `gh pr view --json headRefOid`. When
+# PREFETCH_CONTEXT_ENABLED is off this is a byte-identical no-op (no gh call). If
+# HEAD moved (force-push / new push mid-run), the stale pre-fed context is
+# invalidated and we take the skip sentinel (exit 100) so the scheduler re-runs
+# at the new SHA — never reviewing stale code. The idempotency marker prevents a
+# duplicate review on the retry.
+if [ "${PREFETCH_CONTEXT_ENABLED:-false}" = "true" ]; then
+  _freshness_rc=0
+  assert_prefetch_context_fresh "$PR_URL" "/tmp/cascade" || _freshness_rc=$?
+  if [ "$_freshness_rc" = "100" ]; then
+    echo "    HEAD moved since context prefetch — discarding stale pre-fed context and skipping (will retry at new SHA)"
+    echo "{\"pr\":\"$PR_URL\",\"sha\":\"$PR_HEAD_SHA\",\"decision\":\"skip\",\"reason\":\"head-moved-stale-context\"}"
+    exit 100
+  fi
+  unset _freshness_rc
+fi
+
 # If triage says no concerns → use single-review prompt for quick confirmation
 if [ "$TRIAGE_ESCALATE" = "false" ]; then
   echo "    [approve] triage cleared — running single confirmation ($ENGINE_SINGLE_MODEL)"

--- a/tests/test_prefetch_freshness.bats
+++ b/tests/test_prefetch_freshness.bats
@@ -1,0 +1,190 @@
+#!/usr/bin/env bats
+# Unit tests for the HEAD-SHA freshness safeguard (epic #1101, Story 5 / #1106):
+# assert_prefetch_context_fresh <pr_url> [out_dir].
+#
+# Just before the agentic tiers (deep / audit / single) consume the Story 2
+# pre-fed context, this re-checks the PR's CURRENT head SHA against the SHA
+# stamped on the pre-fed files, so a force-push / new push mid-run can never let
+# a tier review stale code. It must:
+#   - be a no-op returning 0 (and doing NO gh call) when PREFETCH_CONTEXT_ENABLED
+#     is off (AC #4),
+#   - when the current head SHA still MATCHES the stamp, return 0 and leave the
+#     pre-fed context untouched (AC #2 — used as-is, no extra fetch beyond the
+#     one bounded head-SHA call),
+#   - when HEAD MOVED, invalidate the stale pre-fed context (remove the files,
+#     unset the PR_CONTEXT_* exports) and return the skip sentinel 100 so the
+#     caller retries at the new SHA (AC #3),
+#   - bound the check to a SINGLE lightweight `gh pr view --json headRefOid`
+#     (AC #4).
+#
+# `gh` is stubbed via a PATH shim that logs every invocation so we can assert
+# fetch volume and returns a chosen head SHA from STUB_CURRENT_SHA.
+#
+# Run with: bats tests/test_prefetch_freshness.bats
+
+setup() {
+  source "$(dirname "$BATS_TEST_FILENAME")/../scripts/lib/pr-context-prefetch.sh"
+
+  STUB_DIR="$(mktemp -d "$BATS_TEST_TMPDIR/stub.XXXXXX")"
+  export PATH="$STUB_DIR:$PATH"
+  export GH_CALL_LOG="$STUB_DIR/gh_calls.log"
+  : > "$GH_CALL_LOG"
+
+  OUT_DIR="$STUB_DIR/cascade"
+  mkdir -p "$OUT_DIR"
+
+  PR_URL="https://github.com/petry-projects/example/pull/42"
+  STAMPED_SHA="abc123def456"
+}
+
+teardown() {
+  rm -rf "$STUB_DIR"
+}
+
+# Write pre-fed context files stamped with $STAMPED_SHA, mirroring the format
+# prefetch_pr_context persists, and export their paths as the tiers would see.
+_seed_prefed_context() {
+  DIFF_FILE="$OUT_DIR/pr-context-diff.txt"
+  META_FILE="$OUT_DIR/pr-context-metadata.json"
+  printf '# PR_HEAD_SHA: %s\ndiff --git a/x b/x\n' "$STAMPED_SHA" > "$DIFF_FILE"
+  printf '{"number":42,"pr_head_sha":"%s"}' "$STAMPED_SHA" > "$META_FILE"
+  export PR_CONTEXT_DIFF_FILE="$DIFF_FILE"
+  export PR_CONTEXT_METADATA_FILE="$META_FILE"
+}
+
+# gh stub: `pr view` prints {"headRefOid":"$STUB_CURRENT_SHA"} and logs the call.
+_install_gh_head_stub() {
+  cat > "$STUB_DIR/gh" <<'STUB'
+#!/usr/bin/env bash
+echo "$*" >> "$GH_CALL_LOG"
+if [ "$1" = "pr" ] && [ "$2" = "view" ]; then
+  printf '{"headRefOid":"%s"}' "$STUB_CURRENT_SHA"
+  exit 0
+fi
+exit 0
+STUB
+  chmod +x "$STUB_DIR/gh"
+}
+
+# gh stub whose `pr view` fails (e.g. transient API error) on stderr.
+_install_gh_failing_stub() {
+  cat > "$STUB_DIR/gh" <<'STUB'
+#!/usr/bin/env bash
+echo "$*" >> "$GH_CALL_LOG"
+if [ "$1" = "pr" ] && [ "$2" = "view" ]; then
+  echo "could not resolve to a PullRequest" >&2
+  exit 1
+fi
+exit 0
+STUB
+  chmod +x "$STUB_DIR/gh"
+}
+
+# ---------------------------------------------------------------------------
+# Flag OFF: default-off => no gh call, return 0, pre-fed context untouched
+# ---------------------------------------------------------------------------
+@test "freshness OFF (unset): no-op, no gh call" {
+  _install_gh_head_stub
+  export STUB_CURRENT_SHA="ffffffffffff"  # would be a mismatch if it ran
+  _seed_prefed_context
+  unset PREFETCH_CONTEXT_ENABLED
+
+  run assert_prefetch_context_fresh "$PR_URL" "$OUT_DIR"
+  [ "$status" -eq 0 ]
+
+  # No gh call when the feature is off.
+  [ ! -s "$GH_CALL_LOG" ]
+  # Pre-fed context left intact.
+  [ -f "$DIFF_FILE" ]
+  [ -f "$META_FILE" ]
+}
+
+@test "freshness OFF (explicit false): no-op, no gh call" {
+  _install_gh_head_stub
+  export STUB_CURRENT_SHA="ffffffffffff"
+  _seed_prefed_context
+  export PREFETCH_CONTEXT_ENABLED=false
+
+  run assert_prefetch_context_fresh "$PR_URL" "$OUT_DIR"
+  [ "$status" -eq 0 ]
+  [ ! -s "$GH_CALL_LOG" ]
+}
+
+# ---------------------------------------------------------------------------
+# Matched SHA: return 0, exactly one gh pr view, pre-fed context untouched
+# ---------------------------------------------------------------------------
+@test "freshness ON matched: returns 0 and keeps the pre-fed context" {
+  _install_gh_head_stub
+  export STUB_CURRENT_SHA="$STAMPED_SHA"   # HEAD unchanged
+  _seed_prefed_context
+  export PREFETCH_CONTEXT_ENABLED=true
+
+  run assert_prefetch_context_fresh "$PR_URL" "$OUT_DIR"
+  [ "$status" -eq 0 ]
+
+  # Files still present (used as-is).
+  [ -f "$OUT_DIR/pr-context-diff.txt" ]
+  [ -f "$OUT_DIR/pr-context-metadata.json" ]
+
+  # Bounded to a single lightweight head-SHA call.
+  run grep -c 'pr view' "$GH_CALL_LOG"
+  [ "$output" -eq 1 ]
+  # It requested only headRefOid (no full metadata/diff re-fetch).
+  grep -q 'headRefOid' "$GH_CALL_LOG"
+  run grep -c 'pr diff' "$GH_CALL_LOG"
+  [ "$output" -eq 0 ]
+}
+
+# ---------------------------------------------------------------------------
+# Moved SHA: invalidate stale context, unset exports, return 100
+# ---------------------------------------------------------------------------
+@test "freshness ON moved: invalidates stale context and returns 100" {
+  _install_gh_head_stub
+  export STUB_CURRENT_SHA="999999999999"   # force-push moved HEAD
+  _seed_prefed_context
+  export PREFETCH_CONTEXT_ENABLED=true
+
+  # Run in the current shell (not `run`) so we can assert the exports were unset.
+  set +e
+  assert_prefetch_context_fresh "$PR_URL" "$OUT_DIR"
+  rc=$?
+  set -e
+  [ "$rc" -eq 100 ]
+
+  # Stale pre-fed files discarded so no tier can consume them.
+  [ ! -f "$OUT_DIR/pr-context-diff.txt" ]
+  [ ! -f "$OUT_DIR/pr-context-metadata.json" ]
+  # Exports cleared.
+  [ -z "${PR_CONTEXT_DIFF_FILE:-}" ]
+  [ -z "${PR_CONTEXT_METADATA_FILE:-}" ]
+}
+
+# ---------------------------------------------------------------------------
+# Nothing pre-fed: no stamp to validate => no-op (belt-and-suspenders)
+# ---------------------------------------------------------------------------
+@test "freshness ON but nothing pre-fed: no-op returns 0, no gh call" {
+  _install_gh_head_stub
+  export STUB_CURRENT_SHA="$STAMPED_SHA"
+  export PREFETCH_CONTEXT_ENABLED=true
+  unset PR_CONTEXT_DIFF_FILE PR_CONTEXT_METADATA_FILE
+
+  run assert_prefetch_context_fresh "$PR_URL" "$OUT_DIR"
+  [ "$status" -eq 0 ]
+  [ ! -s "$GH_CALL_LOG" ]
+}
+
+# ---------------------------------------------------------------------------
+# gh failure: degrade to proceed (return 0) rather than crash the run
+# ---------------------------------------------------------------------------
+@test "freshness ON gh failure: degrades to 0 and keeps the pre-fed context" {
+  _install_gh_failing_stub
+  _seed_prefed_context
+  export PREFETCH_CONTEXT_ENABLED=true
+
+  run assert_prefetch_context_fresh "$PR_URL" "$OUT_DIR"
+  [ "$status" -eq 0 ]
+
+  # Context untouched — the safeguard must not discard on an inconclusive check.
+  [ -f "$OUT_DIR/pr-context-diff.txt" ]
+  [ -f "$OUT_DIR/pr-context-metadata.json" ]
+}

